MSC4175: Profile field for user time zone (#4175)
matrix-org/matrix-spec-proposals: proposals/4175-profile-field-time-zone.md

# MSC4175: Profile field for user time zone

Knowing another user's time zone is useful for knowing whether they are likely
to respond or not. Example uses include:

* Showing a user's time zone or time zone offset directly.
* Showing a user's local time (with hints of whether it is day or night there).

## Proposal

Profiles can provide an optional `m.tz` field with values equal to names from the
[IANA Time Zone Database](https://www.iana.org/time-zones).

> Clients can set and fetch this via the [normal API endpoints](https://spec.matrix.org/unstable/client-server-api/#profiles).

Reviewer: ```suggestion
Clients can set and fetch this via the [normal API endpoints](https://spec.matrix.org/v1.14/client-server-api/#profiles).
```

(unstable can and does change, breaking links for the future)

* Servers MAY validate that the value is a valid IANA time zone. If deemed invalid
  they MUST return a 400 error with error code `M_INVALID_PARAM`.

> * Clients MUST handle invalid or unknown values.

Reviewer: ```suggestion
* Clients MUST handle invalid or unknown values. One approach may be processing the value as though it was never set.
```

... or something to actually guide implementations on what "MUST" means.

The rationale for somewhat loose validation is that different clients/servers may have
different understanding of valid time zones, e.g. different versions of the time zone
database.

If the field is not provided it SHOULD be interpreted as having no time zone information
for that user.

> 

Reviewer 2: Did you get a chance to do this? Not super important but would be good to resolve the pending stuff.

## Potential issues

Clients may need to understand IANA time zone names to show useful information to users.
Some languages make this easy, e.g. JavaScript can handle this using
[`Date.toLocaleString()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleString).
This may cause clients to bundle the IANA time zone database (and thus also keep it
up to date).

Using the IANA time zone name has the downside that it does now allow arbitrary offsets,
which may be required for time zones which are not internationally recognized.

Clients will need to manually update the profile field when the user changes time zone.
This could be automated by clients based on location, or left as a manual change to
users.

> Clients may wish to periodically fetch the time one of other users as it is

Reviewer 3: ```suggestion
Clients may wish to periodically fetch the time zone of other users as it is
```

liable to change somewhat frequently. Currently, profile data isn't propagated/synchronized
between servers, but that's left to a future MSC to solve. It is recommended that
clients cache the value for 12 - 24 hours.

There should not be backwards compatibility concerns since clients should be ignoring
unknown profile fields.

## Alternatives

The time zone offset could be included directly (in minutes/seconds or in `[+-]HH:MM` form).
This would require clients to manually update the profile field during daylight
savings. Using the IANA time zone name is robust against this.

> ### Delegate profile fields

Reviewer 4: I have to say i am quite uneasy about this MSC - it feels like a slippery slope towards bloating out the Matrix spec with an overlapping and incomplete implementation of JSContact (RFC9553).  

But on the other hand, RFC9553 simply doesn't contain a "current timezone for user" attribute - it's only available as an attribute on physical addresses.  And RFC9553 is for describing addressbook entries... not profile information (which is subtly semantically different: the data you expect to see in your Contacts app in your phone is definitely different shape to the data you expect to see on someone's LinkedIin profile for instance).

So perhaps the best bet here is to just bite the bullet and spec our own field types like this. But I really hope we end up finding a way to put them in their own registry rather than bulk out the rest of the spec - and to indirect to the likes of RFC9553 where we conceivably can.

3 separate fields:

* `tz`: the time zone database name (e.g. `"America/New_York"`)
* `tz_label`: a friendly name (e.g. `"Eastern Daylight Time"`)
* `tz_offset`: offset in seconds as an integer (e.g. `-14400`)

XMPP directly uses either:

* [XEP-054](https://xmpp.org/extensions/xep-0054.html) uses vCard
  ([RFC2426](https://datatracker.ietf.org/doc/html/rfc2426)) converted to XML via
  [draft-dawson-vcard-xml-dtd-01](https://datatracker.ietf.org/doc/html/draft-dawson-vcard-xml-dtd-01)
* [XEP-0292](https://xmpp.org/extensions/xep-0292.html) uses xCard: vCard XML Representation
  ([RFC6351](https://datatracker.ietf.org/doc/html/rfc6351)), see also vCard4
  ([RFC6351](https://datatracker.ietf.org/doc/html/rfc6351))

Rocket.Chat provides a user's [time zone offset](https://developer.rocket.chat/docs/user)
in the `utcOffset` field.

Mattermost [returns an object](https://api.mattermost.com/#tag/users/operation/GetUser)
with the user's manual and/or automatic IANA time zone name.

Discord, Twitter, blah don't provide a user's time zone.

## Security considerations

Showing a user's time zone gives some information to their location. There is currently

> no way to limit what profile fields other users can see.

Author: My assumption was you would just not set it!

I would think clients would already compare to the local clock and not to any information stored in the user's profile.

> no way to limit what profile fields other users can see.

Reviewer 5: Right. I was mainly pointing out that this security consideration is not much of an issue as it can easily just not be set.

If educating users is the concern, we could add a recommendation that clients warn users that the tz will be visible to everyone.

> no way to limit what profile fields other users can see.

Author: Fair enough! I'll expand it a bit to a client recommendation. Thank you!

> no way to limit what profile fields other users can see.

Author: Hopefully I've cleared it up.

# MSC4175: Profile field for user time zone

Knowing another user's time zone is useful for knowing whether they are likely
to respond or not. Example uses include:

* Showing a user's time zone or time zone offset directly.
* Showing a user's local time (with hints of whether it is day or night there).

## Proposal

Profiles can provide an optional `m.tz` field with values equal to names from the
[IANA time zone  database](https://data.iana.org/time-zones/theory.html#naming).
Clients can set and fetch this via the [normal API endpoints](https://spec.matrix.org/unstable/client-server-api/#profiles).

* Servers MAY validate that the value is a valid IANA time zone. If deemed invalid
  they MUST return a 400 error with error code `M_INVALID_PARAM`.
* Clients MUST handle invalid or unknown values.

If the field is not provided it SHOULD be interpreted as having no time zone information
for that user.

## Potential issues

> Clients may need to understand IANA time zone names to show useful information to users.

Reviewer 6: It may be worth nodding to the fact that clients shouldn't cache this value, as it's liable to change (frequently, if tracking the user's location)? I suppose there is a problem in that because this data isn't (loosely) synchronised to a state event you can't really wait for the change to come down sync, so clients should refetch this value upon user interaction?

Author: Added!

* `tz_label`: a friendly name (e.g. `"Eastern Daylight Time"`)
* `tz_offset`: offset in seconds as an integer (e.g. `-14400`)

XMPP directly uses either:

* [XEP-054](https://xmpp.org/extensions/xep-0054.html) uses vCard
  ([RFC2426](https://datatracker.ietf.org/doc/html/rfc2426)) converted to XML via
  [draft-dawson-vcard-xml-dtd-01](https://datatracker.ietf.org/doc/html/draft-dawson-vcard-xml-dtd-01)
* [XEP-0292](https://xmpp.org/extensions/xep-0292.html) uses xCard: vCard XML Representation
  ([RFC6351](https://datatracker.ietf.org/doc/html/rfc6351)), see also vCard4
  ([RFC6351](https://datatracker.ietf.org/doc/html/rfc6351))

Rocket.Chat provides a user's [time zone offset](https://developer.rocket.chat/docs/user)
in the `utcOffset` field.

Mattermost [returns an object](https://api.mattermost.com/#tag/users/operation/GetUser)
with the user's manual and/or automatic IANA time zone name.

Discord, Twitter, blah don't provide a user's time zone.

## Security considerations

Showing a user's time zone gives some information to their location. There is currently
no way to limit what profile fields other users can see.

> clients may wish to warn users when providing a timezone and give

Reviewer 6: ```suggestion
Clients may wish to warn users when providing a timezone and give
```

> clients may wish to warn users when providing a timezone and give
> the option to kot incoude it in their profile.

Reviewer 6: ```suggestion
Clients may wish to warn users when providing a timezone and give
the option to not include it in their profile.
```
